File: Theory and Practice of Computation/prac_2/number_test.java
import java.io.*;

class number {

  static public final int invalid = 0;
  static public final int integral = 1;
  static public final int floating = 2;

  public int tag;
  public int value_i;
  public double value_d;

  public number() { tag = invalid; }
  public number(int v) { tag = integral; value_i = v; }
  public number(double v) { tag = floating; value_d = v; }

  void print() {
    switch (tag) {
    case invalid:
        System.out.print("  <<< invalid number in \"print\" >>> "); break;
    case integral:
        System.out.print(value_i); break;
    case floating:
        System.out.print(value_d); break;
    }
  }

};

class number_test {

  static number plus(number l, number r) {
    number result = new number();

    if(l.tag != r.tag){
      System.out.println("Type Error....!");
    }else{
      result.tag = l.tag;
      if(l.tag==number.integral){
        int res = l.value_i+r.value_i;
        result.value_i = res;
      }else if (r.tag == number.floating){
        double res = l.value_d+r.value_d;
        result.value_d = res;
      }
    }
    return result;
  }

  static public void main(String args[]) {
    number a, b;

    a = new number(5);                              // a is integral
    b = new number(6);                              // so is b

    System.out.print("a = "); a.print(); System.out.print("\n");
    System.out.print("b = "); b.print(); System.out.print("\n");
                                                    // OK: both integral
    System.out.print("a + b = "); plus(a, b).print(); System.out.print("\n");

    a = new number(5.7);                            // a is now floating
    System.out.print("a = "); a.print(); System.out.print("\n");
                                                    // type mismatch expected
    System.out.print("a + b = "); plus(a, b).print(); System.out.print("\n");

    b = new number(3.2);                            // b is now floating
    System.out.print("b = "); b.print(); System.out.print("\n");
                                                    // OK: both floating
    System.out.print("a + b = "); plus(a, b).print(); System.out.print("\n");
  }

}

File: Theory and Practice of Computation/prac/number_test.c
#include <stdio.h>

enum tag_type { invalid, integral, floating };

struct number {
    enum tag_type tag;
    union {
        int i;
        double d;
    } value;
};

void print(struct number n) {
    switch (n.tag) {
    case invalid:
        printf("  <<< invalid number in \"print\" >>> "); break;
    case integral:
        printf("%d", n.value.i); break;
    case floating:
        printf("%g", n.value.d); break;
    }
}

struct number plus (struct number l, struct number r) {
    struct number result; result.tag = invalid;

    if(l.tag!=r.tag){
        printf("Type error....!\n");
    }else{
        result.tag = l.tag;
        if(l.tag==integral){
            int result_num = l.value.i + r.value.i;
            result.value.i = result_num;
        }else{
            double result_num = l.value.d + r.value.d;
            result.value.d = result_num;
        }
    }
    /* more code here */

    return result;
}

int main () {
    struct number a, b;

    a.tag = integral; a.value.i = 5;                // a is integral
    b.tag = integral; b.value.i = 6;                // so is b

    printf("a = "); print(a); printf("\n");
    printf("b = "); print(b); printf("\n");
                                                    // OK: both integral
    printf("a + b = "); print(plus(a, b)); printf("\n");

    a.tag = floating; a.value.d = 5.7;              // a is now floating
    printf("a = "); print(a); printf("\n");
                                                    // type mismatch expected
    printf("a + b = "); print(plus(a, b)); printf("\n");

    b.tag = floating; b.value.d = 3.2;              // b is now floating
    printf("b = "); print(b); printf("\n");
                                                    // OK: both floating
    printf("a + b = "); print(plus(a, b)); printf("\n");

    return 0;
}
